### DISCUSSION TITLE
Scratchaddons and Safari

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#3452] [Discussion]
When Safari will start using a common model with Chrome and Firefox for extensions ([https://www.lifewire.com/soon-chrome-extensions-may-work-in-safari-5188410](https://www.lifewire.com/soon-chrome-extensions-may-work-in-safari-5188410)) will Scratchaddons be able to run on Safari too? If so, is there a tentative date that this is supposed to work?

**Comment by apple502j** [comment]
Date: whenever World_Languages gets enough money to pay the licensing fee.

**Reply by sfederici** [reply]
So, why don't we start a crowd funding?

**Reply by cobaltt7** [reply]
Argentinan tax is weird

**Reply by WorldLanguages** [reply]
For crowd funding, please contact griffpatch directly as he's the real owner of the Scratch Addons project and its placement in extension stores.

**Reply by WorldLanguages** [reply]
I'll ask him what his first impressions are now, and will comment again on this thread when he replies.

**Reply by sfederici** [reply]
Ok. What's the best way to contact him? I'm not sure that -even generally- mentioning extensions in the scratch website would be fine.

**Reply by Secret-chest** [reply]
> Ok. What's the best way to contact him? I'm not sure that -even generally- mentioning extensions in the scratch website would be fine.

@sfederici Use @\griffpatch

**Reply by cobaltt7** [reply]
Or you could contact him on discord

**Reply by OniOkami666** [reply]
I find that making ScratchAddons work on other browsers pretty exciting

**Comment by WorldLanguages** [comment]
On another note, keep in mind compatibility is separate from whether Scratch Addons is on the App Store.

**Reply by sfederici** [reply]
Can you give me some more details?

**Reply by cobaltt7** [reply]
If we had the money to put it on the App Store, we would probably make an effort to do it, even if we had to rewrite parts.

**Reply by WorldLanguages** [reply]
> If we had the money to put it on the App Store, we would do it, even if we had to rewrite the whole thing.

Where did you read that? 🤦 

I meant we could attempt to keep Scratch Addons compatible with Safari, even if we didn't publish on the App Store. There's probably a way to sideload extensions, and if there isn't, someone in the world might come up with a way. It's most likely we'd need at least a few small changes for it to work properly, even if Safari has similar extension APIs.

**Reply by cobaltt7** [reply]
> Where did you read that? 🤦

that's what I interpreted your reply as

I was a bit tired when I wrote that, I rephrased it

**Reply by sfederici** [reply]
> I meant we could attempt to keep Scratch Addons compatible with Safari, even if we didn't publish on the App Store. There's probably a way to sideload extensions, and if there isn't, someone in the world might come up with a way. It's most likely we'd need at least a few small changes for it to work properly, even if Safari has similar extension APIs.

Great. But if the best (or only way) to do it is paying Apple's fees (something that no one is eager to do) I guess that crowd funding could be the solution.

**Reply by GrahamSH-LLK** [reply]
Another plan: Instead of crowdfunding, just have Griffpatch sign us up for https://developer.apple.com/support/membership-fee-waiver/. Voila! 100$ waived.

**Reply by WorldLanguages** [reply]
We're not a non profit organization.

**Reply by CloudGate1234** [reply]
> We're not a non profit organization.

you make money?

**Reply by TheColaber** [reply]
@WorldLanguages means to say that we aren't a legal organization.

**Reply by sfederici** [reply]
> Another plan: Instead of crowdfunding, just have Griffpatch sign us up for https://developer.apple.com/support/membership-fee-waiver/. Voila! 100$ waived.

If the request for a legal organization is not so strict, and no one is interested in getting paid for the development of Scratch Addons (even if, if this is like in Italy, non profit doesn't mean that there shouldn't be anyone that earns money, but only that there are no shares for stock holders) I think this is an excellent idea.

**Reply by BroJac5246** [reply]
> Another plan: Instead of crowdfunding, just have Griffpatch sign us up for https://developer.apple.com/support/membership-fee-waiver/. Voila! 100$ waived.

@WorldLanguages looking back at this, what's the worst that could happen? Our application gets rejected. Obviously we would need someone (with a Mac and programming experience) to look into compatibility, but it makes sense.

**Comment by cobaltt7** [comment]
https://github.com/SA-Userscript/ScratchAddons may work once it's done

**Reply by sfederici** [reply]
> https://github.com/SA-Userscript/ScratchAddons may work once I'm done

Do you mean it could work on safari?

**Reply by cobaltt7** [reply]
ideally

**Reply by sfederici** [reply]
I can't wait to see it!!

**Reply by GrahamSH-LLK** [reply]
> I can't wait to see it!!

Yeah! We're working on fixing settings currently (they will be at https://scratch.mit.edu/sa-extension/settings). You can install (the very broken) userscript from https://userscript.scratchaddons.cf/userscript/script.user.js (use tampermonkey or something similar)

**Reply by sfederici** [reply]
Excellent!! I installed it. Now what is this script supposed to do?

**Reply by CST1229** [reply]
It is supposed to have all the features of Scratch Addons, but in a userscript.

**Reply by cobaltt7** [reply]
pls don't use it yet, it's very broken :p

**Reply by sfederici** [reply]
No worries. I guess this is just the "loader" of the standard ScratchAddons scripts? Are they supposed to work in Safari too?

**Reply by williamhorning** [reply]
> No worries. I guess this is just the "loader" of the standard ScratchAddons scripts? Are they supposed to work in Safari too?

The userscript project is actually just modified ScratchAddons with a loader and some polyfills to make it work as a standard JS script. It should work in Safari if it supports all the JavaScript APIs we use.

**Reply by sfederici** [reply]
Great. Let me know if and when you need any help with testing

**Reply by sfederici** [reply]
> pls don't use it yet, it's very broken :p

Any update?

**Reply by williamhorning** [reply]
@sfederici the SA-Userscript project died a while ago and there aren't any real alternatives afaik but @GrahamSH-LLK would know more than I do.
(also jeez I'm still subscribed to this thread)

**Reply by WorldLanguages** [reply]
We hope to release an official SA userscript at some point. But nothing to announce at the moment.

**Reply by sfederici** [reply]
> We hope to release an official SA userscript at some point. But nothing to announce at the moment.

I see. Thanks

**Reply by cobaltt7** [reply]
I started working on this again this week. I opted to just start over because the old code was very outdated. The settings page appears to work perfectly, now to fix the loader.

**Reply by sfederici** [reply]
> I started working on this again this week. I opted to just start over because the old code was very outdated. The settings page appears to work perfectly, now to fix the loader.

This is great news! Please, let me know if you need testing etc

**Reply by sfederici** [reply]
> I started working on this again this week. I opted to just start over because the old code was very outdated. The settings page appears to work perfectly, now to fix the loader.

Updates?

**Reply by cobaltt7** [reply]
sorry but i've been quite busy lately and haven't been able to get much done on it

**Reply by sfederici** [reply]
Any progress?

**Reply by WorldLanguages** [reply]
No

**Reply by sfederici** [reply]
What a shame!

**Comment by sfederici** [comment]
I don't use much Safari on my mac (mainly because ScratchAddons don't run on it). So, I was surpriesed a few days ago noticing that on the turbowarp.org ScratchAddons' addons -even if limited to the ones focused on the editor- work perfectly. Is this something that hinges on them being natively supported by turbowarp (there is, with no installation) or at least the editor addons could be made available for Scratch too in Safari? If so, why not creating an installable subset of addons limited to the editor for Safari?

**Reply by GrahamSH-LLK** [reply]
The trick is that turbowarp bundles them with the source. The key limitation on Safari is getting anything to run on user devices- extensions have to be signed by apple, with a $100 yearly fee. Various userscript ports have started and then never finished, but if one finally did they would work.

**Reply by sfederici** [reply]
Maybe we could start a crowd funding for this? I could contribute with at least half the amount. But wouldn't be sufficient to ask interested Safari users to run unsigned extensions as explained at https://developer.apple.com/documentation/safariservices/safari_web_extensions/running_your_safari_web_extension? I have created and run a dummy Safari extension on my Mac without owning a developer license.

**Reply by WorldLanguages** [reply]
> But wouldn't be sufficient to ask interested Safari users to run unsigned extensions

Yeah it might be sufficient, but I'm not sure. You need XCode installed, right?

Personally, I don't own a Mac, and I haven't researched what the best way to virtualize MacOS is.

**Reply by sfederici** [reply]
As far I know, you can run unsigned extensions in Safari even if you don't use xcode. Only drawback, you need to manually enable the extension each time you launch Safari:
- Open Safari and choose Safari > Settings
- Select the Advanced tab
- Enable the Show Develop menu in menu bar checkbox
- Click Develop > Allow Unsigned Extensions
- Enter your password and click OK.

To me it would be fine, as I leave Safari open forever. So, if somone has some draft Safari extension, I would gladly test it

**Reply by DNin01** [reply]
> As far I know, you can run unsigned extensions in Safari even if you don't use xcode. Only drawback, you need to manually enable the extension each time you launch Safari: (…)

That's pretty cool for developers, testers, and power users who want to run it in Safari, but if we port Scratch Addons to Safari, it shouldn't require any developer steps for end-users to use it, and we shouldn't recommend that users install it this way.

**Reply by BroJac5246** [reply]
> > As far I know, you can run unsigned extensions in Safari even if you don't use xcode. Only drawback, you need to manually enable the extension each time you launch Safari: (…)
> 
> That's pretty cool for developers, testers, and power users who want to run it in Safari, but if we port Scratch Addons to Safari, it shouldn't require any developer steps for end-users to use it, and we shouldn't recommend that users install it this way.

Not sure how I feel but I just want to put it out there that Kiwi Browser for Android was a bit of a workaround and we still recommended that. It's not a developer step, but it's also not a well-used browser.

**Reply by WorldLanguages** [reply]
> Only drawback, you need to manually enable the extension each time you launch Safari:

That's also the case in Firefox I think, so it's not that bad.

> That's pretty cool for developers, testers, and power users who want to run it in Safari, but if we port Scratch Addons to Safari, it shouldn't require any developer steps for end-users to use it, and we shouldn't recommend that users install it this way.

Yeah, we wouldn't directly recommend users to install the extension in a way where it won't get auto-updates.

**Reply by sfederici** [reply]
> Yeah, we wouldn't directly recommend users to install the extension in a way where it won't get auto-updates.

Correct. But having anyway the possibility, for the ones interested, would be cool.

> Another plan: Instead of crowdfunding, just have Griffpatch sign us up for https://developer.apple.com/support/membership-fee-waiver/. Voila! 100$ waived.

This is really the case for ScratchAddons. Indeed, "You can request to have the annual Apple Developer Program membership fee waived if you’re a **nonprofit organization,** [...] that will **distribute only free apps** on the App Store and is based **in an eligible region.** [...] Eligible regions: Australia, Brazil, Canada, China mainland, France, Germany, Israel, Italy, Japan, Mexico, South Korea, United Kingdom, United States.". So, why not doing it?

**Reply by WorldLanguages** [reply]
I think they mean nonprofit organizations in the traditional sense, legally established organizations.

**Reply by BroJac5246** [reply]
> I think they mean nonprofit organizations in the traditional sense, legally established organizations.

It's worth trying, though, right? The worst that happens is it gets rejected or we can't get all the way through the application form.

Someone (probably griffpatch) could also set up a nonprofit that doesn't do much other than create a legal entity... though that might be a lot of work.

**Reply by WorldLanguages** [reply]
Setting up a nonprofit must cost more than $100 a year :P

**Reply by BroJac5246** [reply]
> Setting up a nonprofit must cost more than $100 a year :P

Still, would it be worth getting in touch and explaining that we're an open-source project that doesn't make profit?

**Reply by DNin01** [reply]
> > Setting up a nonprofit must cost more than $100 a year :P
> 
> Still, would it be worth getting in touch and explaining that we're an open-source project that doesn't make profit?

That's an interesting thought. Maybe it would work :P

**Reply by GrahamSH-LLK** [reply]
> > I think they mean nonprofit organizations in the traditional sense, legally established organizations.
> 
> It's worth trying, though, right? The worst that happens is it gets rejected or we can't get all the way through the application form.

No. They require a DUNS number for a registered nonprofit, you can't even get past the first step. They also will not waive the fee for people who don't have one. In the US at least, creating a nonprofit is expensive and takes a lot of time.

**Reply by GarboMuffin** [reply]
I have offered to publish a Safari version of Scratch Addons on the App Store using my developer account I already pay for (for TurboWarp). WL said "Sounds ok to me". So that's another option instead of the nonprofit waiver.

**Reply by sfederici** [reply]
Thanks a lot @GarboMuffin! So, is there a draft version of ScratchAddons that we can test and see if it can be successfully published on the store?

**Reply by DNin01** [reply]
How about I create a Safari label for issues?

**Reply by DNin01** [reply]
https://github.com/ScratchAddons/ScratchAddons/labels/platform%3A%20safari

**Comment by DNin01** [comment]
Also, I'm not sure if Safari supports Manifest V2.

**Reply by MaterArc** [reply]
As long as it’s safari 15.4 and later, it should!

**Comment by sfederici** [comment]
> Safari supports Manifest V2.

As of Google Bard :-) 

<img width="734" alt="immagine" src="https://github.com/ScratchAddons/ScratchAddons/assets/2651868/bde63390-9f36-4c45-bb92-4c9c71edab65">

**Reply by DNin01** [reply]
Oh.

**Comment by Samq64** [comment]
> Personally, I don't own a Mac, and I haven't researched what the best way to virtualize MacOS is.

I haven't tried it and there might be better options but apparently Virtualbox works after enabling EFI mode and faking parts of the host, but the OS can only be downloaded officially from the app store on an existing Mac.

**Comment by Hans5958** [comment]
Related: #340 (somehow no one links it?)
